File: cardscan/src/main/java/com/getbouncer/cardscan/CreditCard.java
package com.getbouncer.cardscan;

import android.os.Parcel;
import android.os.Parcelable;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.getbouncer.cardscan.base.CreditCardUtils;

public class CreditCard  implements Parcelable {
    @NonNull public final String number;
    @NonNull public final CreditCard.Network network;
    @Nullable public final String expiryMonth;
    @Nullable public final String expiryYear;

    public enum Network {VISA, MASTERCARD, AMEX, DISCOVER, UNIONPAY, UNKNOWN, EGP_MEEZA};

    CreditCard(@NonNull String number, @Nullable String expiryMonth,
                          @Nullable String expiryYear) {
        this.number = number;
        this.expiryMonth = expiryMonth;
        this.expiryYear = expiryYear;

        if (CreditCardUtils.isVisa(number)) {
            this.network = CreditCard.Network.VISA;
        } else if (CreditCardUtils.isAmex(number)) {
            this.network = CreditCard.Network.AMEX;
        } else if (CreditCardUtils.isDiscover(number)) {
            this.network = CreditCard.Network.DISCOVER;
        } else if (CreditCardUtils.isMastercard(number)) {
            this.network = CreditCard.Network.MASTERCARD;
        } else if (CreditCardUtils.isUnionPay(number)) {
            this.network = Network.UNIONPAY;
        } else if (CreditCardUtils.isEgpMeeza(number)) {
            this.network = Network.EGP_MEEZA;
        } else {
            this.network = CreditCard.Network.UNKNOWN;
        }

    }

    @NonNull public String last4() {
        return this.number.substring(this.number.length() - 4);
    }

    @Nullable public String expiryForDisplay() {
        if (this.expiryMonth == null || this.expiryYear == null) {
            return null;
        }

        String month = this.expiryMonth;
        if (month.length() == 1) {
            month = "0" + month;
        }
        String year = this.expiryYear;
        if (year.length() == 4) {
            year = year.substring(2);
        }

        return month + "/" + year;
    }

    private CreditCard(Parcel in) {
        String number = in.readString();
        this.expiryMonth = in.readString();
        this.expiryYear = in.readString();
        CreditCard.Network network = (CreditCard.Network) in.readSerializable();

        if (number == null) {
            // this should never happen, but makes the compiler happy
            this.number = "";
        } else {
            this.number = number;
        }

        if (network == null) {
            this.network = CreditCard.Network.UNKNOWN;
        } else {
            this.network = network;
        }
    }

    public static final Creator<CreditCard> CREATOR = new Creator<CreditCard>() {
        @Override
        public CreditCard createFromParcel(Parcel in) {
            return new CreditCard(in);
        }

        @Override
        public CreditCard[] newArray(int size) {
            return new CreditCard[size];
        }
    };

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel parcel, int i) {
        parcel.writeString(number);
        parcel.writeString(expiryMonth);
        parcel.writeString(expiryYear);
        parcel.writeSerializable(network);
    }
}

File: cardscan/src/main/java/com/getbouncer/cardscan/CreditCard.kt
package com.getbouncer.cardscan

import android.os.Parcel
import android.os.Parcelable
import com.getbouncer.cardscan.base.CreditCardUtils
import com.getbouncer.cardscan.base.CardNetwork

class CreditCard : Parcelable {
    val number: String
    val network: CardNetwork
    val expiryMonth: String?
    val expiryYear: String?

    internal constructor(number: String, expiryMonth: String?, expiryYear: String?) {
        this.number = number
        this.expiryMonth = expiryMonth
        this.expiryYear = expiryYear
        this.network = CreditCardUtils.determineCardNetwork(number)
    }

    fun last4(): String = this.number.substring(this.number.length - 4)

    fun expiryForDisplay(): String? =
        CreditCardUtils.formatExpirationForDisplay(this.expiryMonth, this.expiryYear)

    fun numberForDisplay(): String = CreditCardUtils.formatNumberForDisplay(number)

    private constructor(parcel: Parcel) {
        val number = parcel.readString()
        this.expiryMonth = parcel.readString()
        this.expiryYear = parcel.readString()
        val network = parcel.readSerializable() as CardNetwork?

        if (number == null) {
            // this should never happen, but makes the compiler happy
            this.number = ""
        } else {
            this.number = number
        }

        if (network == null) {
            this.network = CardNetwork.UNKNOWN
        } else {
            this.network = network
        }
    }

    override fun describeContents(): Int = 0

    override fun writeToParcel(parcel: Parcel, i: Int) {
        parcel.writeString(number)
        parcel.writeString(expiryMonth)
        parcel.writeString(expiryYear)
        parcel.writeSerializable(network)
    }

    companion object CREATOR : Parcelable.Creator<CreditCard> {
        override fun createFromParcel(parcel: Parcel): CreditCard = CreditCard(parcel)
        override fun newArray(size: Int): Array<CreditCard?> = arrayOfNulls(size)
    }
}